Sources/NIOHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,333 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIOHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Swift Server Working Group and the SwiftNIOHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIOHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOHTTP2
+import NIOTLS
+
+class ConnectionPool {
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.eventLoopGroup = group
+        self.configuration = configuration
+    }
+
+    let eventLoopGroup: EventLoopGroup
+    let configuration: HTTPClient.Configuration
+    private let lock = Lock()
+
+    private var _connectionProviders: [Key: ConnectionProvider] = [:]
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    func getConnection(for request: HTTPClient.Request) -> EventLoopFuture<Connection> {
+        
+        let key: Key
+        do {
+            key = try Key(url: request.url)
+        } catch  {
+            return eventLoopGroup.next().makeFailedFuture(error)
+        }
+
+        //FIXME: See what to do with timeout
```

Comment:
what's wrong with:
```swift
if let readTimeout = timeout.read {
    return channel.pipeline.addHandler(IdleStateHandler(readTimeout: readTimeout))
} else {
    return channel.eventLoop.makeSucceededFuture(())
}
```
I think we can try to remove and re-add `IdleHandler` if timeout is different. Alternatively, timeout can be part of the `Key`

Reply:
Since the pool consumer is already the one that adds its own `TaskHandler` I wondered if it wouldn't be best for the connection pool not to add the `IdleStateHandler` associated with `configuration.timeout either`.

I think it will still add an `IdleStateHandler`, but only for its unleased connections, based on a default timeout delay or on what was set in the `Keep-Alive` header, if available.

Does this seem correct to you?